### PR TITLE
Convert instrumentation reports to chrome tracing format and improve running of SDFGs

### PR DESCRIPTION
**Pull request by phschaad**
- [x] Convert instrumentation reports to use the chrome tracing format
- [x] Enable SDFG-level instrumentation
- [x] Add a CLI tool to summarize instrumentation reports on the console
- [x] Move all dace CLI tools to a unified dace CLI module
- [x] Save links between generating Python code and generated SDFGs to facilitate running SDFGs more easily

### CODE REVIEW

File: dace/codegen/codegen.py [hunk 1]
```diff
@@ -134,6 +134,9 @@ def generate_code(sdfg) -> List[CodeObject]:
         elif hasattr(node, 'map'):
             frame._dispatcher.instrumentation[node.map.instrument] = \
                 provider_mapping[node.map.instrument]
+    if hasattr(sdfg, 'instrument'):
+        frame._dispatcher.instrumentation[sdfg.instrument] = \
+            provider_mapping[sdfg.instrument]
```
Comment (tbennun) [inline dace/codegen/codegen.py:139]: shouldn't this be just `if sdfg.instrument != dtypes.InstrumentationType.No_Instrumentation`?

File: dace/frontend/python/parser.py [hunk 2]
```diff
@@ -129,8 +129,30 @@ def parse_from_function(function, *compilation_args, strict=None):
...
+                sdfg.name + ',' +
+                os.path.abspath(os.path.join(sdfg.build_folder, 'program.sdfg'))
+                + ',' + os.path.abspath(sdfg_out_path) + ',' +
+                os.path.abspath(function.cli_args[0]) + ',' +
+                ','.join([str(el) for el in function.cli_args])
+            )
+
+    # Save the SDFG (again). Skip this step if running from a cached SDFG, as
+    # it might overwrite the cached SDFG.
+    if Config.get_bool('compiler', 'use_cache'):
```
Comment (tbennun) [inline dace/frontend/python/parser.py:154]: this if condition seems inverted

File: dace/frontend/python/parser.py [hunk 3]
```diff
@@ -129,8 +129,30 @@ def parse_from_function(function, *compilation_args, strict=None):
...
+                    ['argv_' + str(i) for i in range(len(function.cli_args))]
+                ) + '\n'
+            )
+            launchfiles_file.write(
+                sdfg.name + ',' +
+                os.path.abspath(os.path.join(sdfg.build_folder, 'program.sdfg'))
+                + ',' + os.path.abspath(sdfg_out_path) + ',' +
+                os.path.abspath(function.cli_args[0]) + ',' +
+                ','.join([str(el) for el in function.cli_args])
+            )
```
Comment (tbennun) [inline dace/frontend/python/parser.py:150]: Why is this specific to the Python frontend?

File: dace/frontend/python/parser.py [hunk 4]
```diff
@@ -232,12 +254,13 @@ def infer_symbols_from_shapes(sdfg: SDFG, args: Dict[str, Any],
 class DaceProgram:
     """ A data-centric program object, obtained by decorating a function with
         `@dace.program`. """
-    def __init__(self, f, args, kwargs):
+    def __init__(self, f, args, kwargs, cli_args):
```
Comment (tbennun) [inline dace/frontend/python/parser.py:257]: why specifically in the python frontend?

File: dace/runtime/include/dace/perf/reporting.h [hunk 5]
```diff
@@ -9,73 +9,185 @@
...
+
+#define DACE_REPORT_BUFFER_SIZE 2048
+
 namespace dace {
 namespace perf {
 
+    struct TraceEvent {
+        char ph;
+        std::string name;
+        std::string cat;
```
Comment (tbennun) [inline dace/runtime/include/dace/perf/reporting.h:28]: wouldn't dynamically-sized strings be slow?

File: dace/runtime/include/dace/perf/reporting.h [hunk 6]
```diff
@@ -9,73 +9,185 @@
...
+        unsigned long int tstart;
+        unsigned long int tend;
+        std::thread::id tid;
+        struct _element_id {
+            int sdfg_id;
+            int state_id;
+            int el_id;
+        } element_id;
+        struct _counter {
+            std::string name;
```
Comment (tbennun) [inline dace/runtime/include/dace/perf/reporting.h:38]: same

File: dace/sdfg/sdfg.py [hunk 7]
```diff
@@ -363,21 +370,36 @@ def from_json(cls, json_obj, context_info=None):
 
         return ret
 
-    def hash_sdfg(self) -> int:
+    def hash_sdfg(self) -> str:
         '''
         Returns an hash of the current SDFG, without considering IDs and attributes name
```
Comment (tbennun) [inline dace/sdfg/sdfg.py:375]:
```suggestion
        Returns a hash of the current SDFG, without considering IDs and attribute names.
```

File: dace/sdfg/sdfg.py [hunk 8]
```diff
@@ -363,21 +370,36 @@ def from_json(cls, json_obj, context_info=None):
...
         :return: the hash
```
Comment (tbennun) [inline dace/sdfg/sdfg.py:376]:
```suggestion
        :return: The hash (in SHA-256 format).
```

File: dace/sdfg/sdfg.py [hunk 9]
```diff
@@ -386,8 +408,9 @@ def keyword_remover(json_obj: Any, last_keyword=""):
         jsondict = self.to_json()  # No more nonstandard objects
         keyword_remover(jsondict)  # Make non-unique in SDFG hierarchy
         string_representation = dace.serialize.dumps(jsondict)  # dict->str
-        hsh = hash(string_representation)  # str->int
-        return hsh
+        hsh = sha256(string_representation.encode('utf-8'))
+        hsh.update(string_representation.encode('utf-8'))
```
Comment (tbennun) [inline dace/sdfg/sdfg.py:412]: why two updates with the same string?

**Review by tbennun — CHANGES_REQUESTED** [review]
Some questions and changes.

File: dace/codegen/codegen.py (hunk 1 above)
Comment (phschaad) [reply dace/codegen/codegen.py]: That's true, addressed.

File: dace/frontend/python/parser.py (hunk 2 above)
Comment (phschaad) [reply dace/frontend/python/parser.py]: Good catch, thanks! Simplification gone wrong..

File: dace/frontend/python/parser.py (hunk 3 above)
Comment (phschaad) [reply dace/frontend/python/parser.py]: Moved it to be a feature of all SDFGs.

File: dace/frontend/python/parser.py (hunk 4 above)
Comment (phschaad) [reply dace/frontend/python/parser.py]: dito

File: dace/sdfg/sdfg.py (hunk 9 above)
Comment (phschaad) [reply dace/sdfg/sdfg.py]: That's a mistake, good catch

File: dace/runtime/include/dace/perf/reporting.h (hunk 5 above)
Comment (phschaad) [reply dace/runtime/include/dace/perf/reporting.h]: Uses fixed char array buffering now, seems to indeed be minutely faster.

**Review by tbennun — CHANGES_REQUESTED** [review]
Looks good, very minor comments!

**Review by tbennun — APPROVED** [review]
LGTM! One of the cooler PRs recently.

**Pull request merged**